File: 2023/java/src/main/java/org/frieder/aoc/day5/b/ProcessRangeTask.java
package org.frieder.aoc.day5.b;

import lombok.RequiredArgsConstructor;

import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

import static org.frieder.aoc.day5.b.Solution5B.SEEK;

@RequiredArgsConstructor
public class ProcessRangeTask {

    private Double currentLocation;
    private final Seed seed;
    private final Map<String, Data> data;

    public Double getLocationNumber() {
        for (double j = 0; j < this.seed.getRange(); j++) {
            double seedNumber = this.seed.getStart() + j;
            String pathId = this.searchSeed(seedNumber);
            if (j + SEEK >= this.seed.getRange())
                continue;

            j += (pathId.equals(this.searchSeed(seedNumber + SEEK))) ? SEEK : 0;
        }

        return this.currentLocation;
    }

    private String searchSeed(Double seedNumber) {

        TargetResult r = this.getTarget("seed", "location", seedNumber, "");
        if (this.currentLocation == null) {
            this.currentLocation = r.getLocationNumber();
        } else {
            this.currentLocation = Math.min(this.currentLocation, r.getLocationNumber());
        }
        return r.getPathId();
    }

    private TargetResult getTarget(String source, String target, Double initial, String pathId) {

        if (target.equals(source))
            return new TargetResult(null, null);

        AtomicReference<String> hit = new AtomicReference<>();
        Double destination = this.data.get(source).getMappings()
                .stream()
                .filter(m -> initial >= m.getSourceStart() && initial < m.getSourceStart() + m.getRange())
                .findFirst()
                .map(m -> {
                    hit.set(m.getSourceStart() + ":" + m.getDestinationStart());
                    return m.getDestinationStart() + (initial - m.getSourceStart());
                })
                .orElse(initial);

        if (target.equals(this.data.get(source).getTarget())) {
            return new TargetResult(
                    destination,
                    pathId + source + "-to-" + this.data.get(source).getTarget() + ":" + hit.get() + ">"
            );
        }

        return this.getTarget(this.data.get(source).getTarget(), target, destination, pathId);
    }


}

File: 2023/php/src/day5/b/ProcessRangeTask.php
<?php

use Amp\Cancellation;
use Amp\Parallel\Worker\Task;
use Amp\Sync\Channel;

class ProcessRangeTask implements Task
{
    private mixed $curLoc = null;

    public function __construct(
        private readonly string $sourceStart,
        private readonly string $range,
        private readonly array  $data,
        private readonly int    $seek

    )
    {
    }

    private function searchSeed(int $seed): string
    {
        $r = self::getTarget("seed", "location", $seed);
        if ($this->curLoc == null) {
            $this->curLoc = $r["destination"];
        } else {
            $this->curLoc = min($this->curLoc, $r["destination"]);
        }
        return $r["pathid"];
    }

    public function run(Channel $channel, Cancellation $cancellation): mixed
    {

        for ($j = 0; $j < intval($this->range); $j++) {
            $seed = intval($this->sourceStart) + $j;
            $pathid = $this->searchSeed($seed);

            if ($j + $this->seek >= intval($this->range))
                continue;

            $j += ($pathid == self::searchSeed($seed + $this->seek)) ? $this->seek : 0;
        }

        return $this->curLoc;
    }


    private function getTarget(string $source, string $target, int $initial, $pathid = ""): array
    {
        if ($source == $target)
            return array();

        $destination = $initial;
        $hit = "-1";

        foreach ($this->data[$source]["mapping"] as $i => $mapping) {
            if ($initial >= $mapping["sourceStart"] && $initial < $mapping["sourceStart"] + $mapping["range"]) {
                $destination = $mapping["destinationStart"] + ($initial - $mapping["sourceStart"]);
                $hit = $i;
                break;
            }
        }

        if ($this->data[$source]["target"] == $target) {
            return array(
                "destination" => $destination,
                "pathid" => $pathid . $source . "-to-" . $this->data[$source]["target"] . ":" . $hit . ">"
            );
        }

        return self::getTarget($this->data[$source]["target"], $target, $destination, $pathid);
    }
}
